tls/s2n_client_hello.c

```diff
@@ -603,3 +603,50 @@ int s2n_client_hello_get_session_id(struct s2n_client_hello *ch, uint8_t *out, u
 
     return S2N_SUCCESS;
 }
+
+static S2N_RESULT s2n_client_hello_get_raw_extension(uint16_t expected_extension_type,
+        struct s2n_blob *raw_extensions, struct s2n_blob *extension)
+{
+    RESULT_GUARD_PTR(raw_extensions);
+
+    struct s2n_stuffer raw_extensions_stuffer = { 0 };
+    RESULT_GUARD_POSIX(s2n_stuffer_init(&raw_extensions_stuffer, raw_extensions));
+    RESULT_GUARD_POSIX(s2n_stuffer_skip_write(&raw_extensions_stuffer, raw_extensions->size));
+
+    while(s2n_stuffer_data_available(&raw_extensions_stuffer) > 0) {
+        uint16_t extension_type = 0;
+        RESULT_GUARD_POSIX(s2n_stuffer_read_uint16(&raw_extensions_stuffer, &extension_type));
+
+        uint16_t extension_size = 0;
+        RESULT_GUARD_POSIX(s2n_stuffer_read_uint16(&raw_extensions_stuffer, &extension_size));
+
+        uint8_t *extension_data = s2n_stuffer_raw_read(&raw_extensions_stuffer, extension_size);
+        RESULT_ENSURE_REF(extension_data);
+
+        if (expected_extension_type == extension_type) {
+            RESULT_GUARD_POSIX(s2n_blob_init(extension, extension_data, extension_size));
+            return S2N_RESULT_OK;
+        }
+    }
+    return S2N_RESULT_ERROR;
```

Comment:
We should never return failure without setting s2n_errno. You want to either return success or RESULT_BAIL with a specific error.

Reply:
Like you suggested, going to fail with error S2N_ERR_INVALID_ARGUMENT